test-runner/src/main/scala/scala/scalanative/testinterface/ComRunner.scala

```diff
@@ -56,51 +55,44 @@ class ComRunner(bin: File,
     new BufferedOutputStream(socket.getOutputStream))
 
   /** Send message `msg` to the distant program. */
-  def send(msg: Message): Unit = synchronized {
-    try SerializedOutputStream(out)(_.writeMessage(msg))
-    catch {
-      case ex: Throwable =>
-        close()
-        throw ex
+  def send(msg: Message): Unit = {
+    synchronized { // Here, not at def, to workaround SN Issue #1091.
+      try SerializedOutputStream(out)(_.writeMessage(msg))
+      catch {
+        case ex: Throwable =>
+          close()
+          throw ex
+      }
     }
   }
 
   /** Wait for a message to arrive from the distant program. */
-  def receive(timeout: Duration = Duration.Inf): Message =
-    synchronized {
-      in.mark(Int.MaxValue)
-      val savedSoTimeout = socket.getSoTimeout()
-      try {
-        val deadLineMs = if (timeout.isFinite()) timeout.toMillis else 0L
-        socket.setSoTimeout((deadLineMs min Int.MaxValue).toInt)
-
-        val result =
-          SerializedInputStream.next(in)(_.readMessage()) match {
-            case logMsg: Log =>
-              log(logMsg)
-              receive(timeout)
-            case other =>
-              other
-          }
-
-        in.mark(0)
-        result
+  def receive(): Message = {
+    @tailrec
+    def loop(): Message = {
```

Comment:
Please define this in the narrowest possible scope, i.e., within the `try { ... }` block. That way it's crystal clear that the code of `loop()` is always protected by the `try`.

Reply:
Understood & done. Good idea, thank you.